// @ziplayer/extension/dist/index.js
"use strict";
/**
 * @fileoverview Main export file for ZiPlayer extensions.
 *
 * This module exports all available extensions and supporting classes for the ZiPlayer
 * music bot framework. Extensions provide additional functionality beyond basic
 * audio playback, such as Lavalink integration, voice recognition, and lyrics fetching.
 *
 * @example
 * ```typescript
 * import { lavalinkExt, voiceExt, lyricsExt, NodeManager } from "ziplayer/extension";
 *
 * const manager = new PlayerManager({
 *   extensions: [
 *     new lavalinkExt(null, {
 *       nodes: [{ host: "localhost", port: 2333, password: "youshallnotpass" }]
 *     }),
 *     new voiceExt(null, { lang: "en-US" }),
 *     new lyricsExt(null, { provider: "lrclib" })
 *   ]
 * });
 * ```
 *
 * @since 1.0.0
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TrackResolver = exports.VoiceHandler = exports.WebSocketHandler = exports.PlayerStateManager = exports.NodeManager = exports.lyricsExt = exports.voiceExt = exports.lavalinkExt = void 0;
/**
 * Lavalink extension for high-performance audio streaming.
 *
 * Provides integration with Lavalink nodes for:
 * - High-quality audio streaming with low latency
 * - Advanced audio processing and effects
 * - Load balancing across multiple nodes
 * - WebSocket-based real-time updates
 *
 * @example
 * ```typescript
 * const lavalinkExt = new lavalinkExt(null, {
 *   nodes: [{ host: "localhost", port: 2333, password: "youshallnotpass" }]
 * });
 * ```
 */
var lavalinkExt_1 = require("./lavalinkExt");
Object.defineProperty(exports, "lavalinkExt", { enumerable: true, get: function () { return lavalinkExt_1.lavalinkExt; } });
/**
 * Voice extension for real-time speech recognition.
 *
 * Provides voice-to-text functionality including:
 * - Real-time speech recognition in Discord voice channels
 * - Multiple language support
 * - Custom speech resolver support
 * - Audio quality filtering
 *
 * @example
 * ```typescript
 * const voiceExt = new voiceExt(null, {
 *   lang: "en-US",
 *   ignoreBots: true
 * });
 * ```
 */
var voiceExt_1 = require("./voiceExt");
Object.defineProperty(exports, "voiceExt", { enumerable: true, get: function () { return voiceExt_1.voiceExt; } });
/**
 * Lyrics extension for automatic lyrics fetching and synchronization.
 *
 * Provides lyrics functionality including:
 * - Automatic lyrics fetching from multiple providers
 * - LRC (synchronized) lyrics support
 * - Real-time line-by-line updates
 * - Title sanitization for better matching
 *
 * @example
 * ```typescript
 * const lyricsExt = new lyricsExt(null, {
 *   provider: "lrclib",
 *   includeSynced: true
 * });
 * ```
 */
var lyricsExt_1 = require("./lyricsExt");
Object.defineProperty(exports, "lyricsExt", { enumerable: true, get: function () { return lyricsExt_1.lyricsExt; } });
/**
 * NodeManager for managing Lavalink node connections.
 *
 * Handles:
 * - Connection management to multiple Lavalink nodes
 * - Load balancing and node selection
 * - REST API operations
 * - WebSocket event handling
 *
 * @example
 * ```typescript
 * const nodeManager = new NodeManager({
 *   nodes: [{ host: "localhost", port: 2333, password: "youshallnotpass" }]
 * });
 * ```
 */
var NodeManager_1 = require("./managers/NodeManager");
Object.defineProperty(exports, "NodeManager", { enumerable: true, get: function () { return NodeManager_1.NodeManager; } });
/**
 * PlayerStateManager for managing player states and voice connections.
 *
 * Handles:
 * - Player state tracking and synchronization
 * - Voice connection state management
 * - Voice server and state update handling
 * - Player-to-node mapping
 *
 * @example
 * ```typescript
 * const stateManager = new PlayerStateManager(true);
 * stateManager.attachPlayer(player);
 * ```
 */
var PlayerStateManager_1 = require("./managers/PlayerStateManager");
Object.defineProperty(exports, "PlayerStateManager", { enumerable: true, get: function () { return PlayerStateManager_1.PlayerStateManager; } });
/**
 * WebSocketHandler for managing WebSocket connections to Lavalink nodes.
 *
 * Handles:
 * - WebSocket connection management
 * - Event handling and routing
 * - Automatic reconnection
 * - Connection state tracking
 */
var WebSocketHandler_1 = require("./handlers/WebSocketHandler");
Object.defineProperty(exports, "WebSocketHandler", { enumerable: true, get: function () { return WebSocketHandler_1.WebSocketHandler; } });
/**
 * VoiceHandler for managing voice connection operations.
 *
 * Handles:
 * - Voice connection establishment
 * - Voice server and state updates
 * - Gateway payload sending
 * - Connection state management
 */
var VoiceHandler_1 = require("./handlers/VoiceHandler");
Object.defineProperty(exports, "VoiceHandler", { enumerable: true, get: function () { return VoiceHandler_1.VoiceHandler; } });
/**
 * TrackResolver for resolving and encoding tracks for Lavalink.
 *
 * Handles:
 * - Track resolution from various sources
 * - Track encoding for Lavalink compatibility
 * - Search functionality
 * - Track metadata processing
 */
var TrackResolver_1 = require("./resolvers/TrackResolver");
Object.defineProperty(exports, "TrackResolver", { enumerable: true, get: function () { return TrackResolver_1.TrackResolver; } });
/**
 * Utility functions and helpers for extensions.
 *
 * Provides:
 * - Debug logging utilities
 * - Track encoding/decoding helpers
 * - Common utility functions
 * - Type checking helpers
 */
__exportStar(require("./utils/helpers"), exports);
/**
 * Type definitions for Lavalink integration.
 *
 * Includes:
 * - Lavalink node configuration types
 * - Player state types
 * - Voice connection types
 * - Extension option types
 */
__exportStar(require("./types/lavalink"), exports);
