// ziplayer/dist/plugins/YouTubePlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.YouTubePlugin = void 0;
const BasePlugin_1 = require("./BasePlugin");
const youtubei_js_1 = require("youtubei.js");
class YouTubePlugin extends BasePlugin_1.BasePlugin {
    constructor() {
        super();
        this.name = "youtube";
        this.version = "1.0.0";
        this.ready = this.init();
    }
    async init() {
        this.client = await youtubei_js_1.Innertube.create({
            client_type: "ANDROID",
            retrieve_player: false,
        });
        // Use a separate web client for search to avoid mobile parser issues
        this.searchClient = await youtubei_js_1.Innertube.create({
            client_type: "WEB",
            retrieve_player: false,
        });
        youtubei_js_1.Log.setLevel(0);
    }
    canHandle(query) {
        const q = query.toLowerCase();
        const isUrl = q.startsWith("http://") || q.startsWith("https://");
        return (q.includes("youtube.com") ||
            q.includes("youtu.be") ||
            (!isUrl && q.includes("youtube")));
    }
    validate(url) {
        const u = url.toLowerCase();
        return u.includes("youtube.com") || u.includes("youtu.be");
    }
    async search(query, requestedBy) {
        await this.ready;
        if (this.validate(query)) {
            const listId = this.extractListId(query);
            if (listId) {
                try {
                    const playlist = await this.client.getPlaylist(listId);
                    const videos = playlist?.videos || playlist?.items || [];
                    const tracks = videos.map((v) => {
                        const id = v.id || v.video_id || v.videoId;
                        const title = v.title?.text ?? v.title;
                        const duration = toSeconds(v.duration?.text ?? v.duration);
                        const thumb = v.thumbnails?.[0]?.url || v.thumbnail?.url;
                        const author = v.author?.name ?? v.channel?.name;
                        const views = v.view_count ?? v.views;
                        return {
                            id: String(id),
                            title,
                            url: `https://www.youtube.com/watch?v=${id}`,
                            duration: Number(duration) || 0,
                            thumbnail: thumb,
                            requestedBy,
                            source: this.name,
                            metadata: { author, views, playlist: listId },
                        };
                    });
                    return {
                        tracks,
                        playlist: {
                            name: playlist?.title ||
                                playlist?.metadata?.title ||
                                `Playlist ${listId}`,
                            url: query,
                            thumbnail: playlist?.thumbnails?.[0]?.url || playlist?.thumbnail?.url,
                        },
                    };
                }
                catch {
                    const withoutList = query
                        .replace(/[?&]list=[^&]+/, "")
                        .replace(/[?&]$/, "");
                    return await this.search(withoutList, requestedBy);
                }
            }
            const videoId = this.extractVideoId(query);
            if (!videoId)
                throw new Error("Invalid YouTube URL");
            const info = await this.client.getBasicInfo(videoId);
            const basic = info.basic_info ?? {};
            const track = {
                id: videoId,
                title: basic.title ?? info.title ?? "Unknown title",
                url: `https://www.youtube.com/watch?v=${videoId}`,
                duration: toSeconds(basic.duration ?? info.duration) ?? 0,
                thumbnail: basic.thumbnail?.[0]?.url ||
                    basic.thumbnail?.[basic.thumbnail?.length - 1]?.url ||
                    info.thumbnails?.[0]?.url,
                requestedBy,
                source: this.name,
                metadata: {
                    author: basic.author ?? info.author?.name,
                    views: info.basic_info?.view_count ?? info.view_count,
                },
            };
            return { tracks: [track] };
        }
        // Text search → return up to 10 video tracks
        const res = await this.searchClient.search(query, {
            type: "video",
        });
        const items = res?.items || res?.videos || res?.results || [];
        const tracks = items.slice(0, 10).map((v) => {
            const id = v.id || v.video_id || v.videoId || v.identifier;
            const title = v.title?.text ?? v.title ?? v.headline ?? "Unknown title";
            const duration = toSeconds(v.duration?.text ?? v.duration?.seconds ?? v.duration ?? v.length_text);
            const thumbnail = v.thumbnails?.[0]?.url ||
                v.thumbnail?.url ||
                v.thumbnail?.thumbnails?.[0]?.url;
            const author = v.author?.name ?? v.channel?.name ?? v.owner?.name;
            const views = v.view_count ?? v.views ?? v.short_view_count ?? v.stats?.view_count;
            const track = {
                id: String(id),
                title,
                url: `https://www.youtube.com/watch?v=${id}`,
                duration: Number(duration) || 0,
                thumbnail,
                requestedBy,
                source: this.name,
                metadata: { author, views },
            };
            return track;
        });
        return { tracks };
    }
    async extractPlaylist(url, requestedBy) {
        await this.ready;
        const listId = this.extractListId(url);
        if (!listId)
            return [];
        try {
            const playlist = await this.client.getPlaylist(listId);
            const videos = playlist?.videos || playlist?.items || [];
            return videos.map((v) => {
                const id = v.id || v.video_id || v.videoId;
                const title = v.title?.text ?? v.title;
                const duration = toSeconds(v.duration?.text ?? v.duration);
                const thumb = v.thumbnails?.[0]?.url || v.thumbnail?.url;
                const author = v.author?.name ?? v.channel?.name;
                const views = v.view_count ?? v.views;
                const track = {
                    id: String(id),
                    title,
                    url: `https://www.youtube.com/watch?v=${id}`,
                    duration: Number(duration) || 0,
                    thumbnail: thumb,
                    requestedBy,
                    source: this.name,
                    metadata: { author, views, playlist: listId },
                };
                return track;
            });
        }
        catch {
            // If playlist fetch fails, return empty to keep optional contract intact
            return [];
        }
    }
    async getStream(track) {
        await this.ready;
        const id = this.extractVideoId(track.url) || track.id;
        if (!id)
            throw new Error("Invalid track id");
        try {
            const stream = await this.client.download(id, {
                type: "audio",
                quality: "best",
            });
            return {
                stream,
                type: "arbitrary",
                metadata: track.metadata,
            };
        }
        catch (e) {
            try {
                const info = await this.client.getBasicInfo(id);
                // Prefer m4a audio-only formats first
                let format = info?.chooseFormat?.({
                    type: "audio",
                    quality: "best",
                });
                if (!format && info?.formats?.length) {
                    const audioOnly = info.formats.filter((f) => f.mime_type?.includes("audio"));
                    audioOnly.sort((a, b) => (b.bitrate || 0) - (a.bitrate || 0));
                    format = audioOnly[0];
                }
                if (!format)
                    throw new Error("No audio format available");
                let url = undefined;
                if (typeof format.decipher === "function") {
                    url = format.decipher(this.client.session.player);
                }
                if (!url)
                    url = format.url;
                if (!url)
                    throw new Error("No valid URL to decipher");
                const res = await fetch(url);
                if (!res.ok || !res.body) {
                    throw new Error(`HTTP ${res.status}`);
                }
                return {
                    stream: res.body,
                    type: "arbitrary",
                    metadata: {
                        ...track.metadata,
                        itag: format.itag,
                        mime: format.mime_type,
                    },
                };
            }
            catch (inner) {
                throw new Error(`Failed to get YouTube stream: ${inner?.message || inner}`);
            }
        }
    }
    async getRelatedTracks(trackURL, opts = {}) {
        await this.ready;
        const videoId = this.extractVideoId(trackURL);
        const info = await await this.searchClient.getInfo(videoId);
        const related = info?.watch_next_feed || [];
        const offset = opts.offset ?? 0;
        const limit = opts.limit ?? 5;
        const relatedfilter = related.filter((tr) => !(opts?.history ?? []).some((t) => t.url === tr.url));
        return relatedfilter.slice(offset, offset + limit).map((v) => {
            const id = v.id || v.video_id || v.videoId || v.content_id;
            const videometa = v?.metadata;
            return {
                id: String(id),
                title: videometa.title.text ?? "Unknown title",
                url: `https://www.youtube.com/watch?v=${id}`,
                duration: Number(v.length_seconds || toSeconds(v.duration)) || 0,
                thumbnail: v.thumbnails?.[0]?.url ||
                    v.thumbnail?.url ||
                    v.content_image?.image[0]?.url,
                requestedBy: "auto",
                source: this.name,
                metadata: { author: v.author, views: v.view_count },
            };
        });
    }
    async getFallback(track) {
        try {
            const result = await this.search(track.title, track.requestedBy);
            const first = result.tracks[0];
            if (!first)
                throw new Error("No fallback track found");
            return await this.getStream(first);
        }
        catch (e) {
            throw new Error(`YouTube fallback search failed: ${e?.message || e}`);
        }
    }
    extractVideoId(input) {
        try {
            const u = new URL(input);
            if (u.hostname.includes("youtu.be")) {
                return u.pathname.split("/").filter(Boolean)[0] || null;
            }
            if (u.hostname.includes("youtube.com")) {
                // watch?v=, shorts/, embed/
                if (u.searchParams.get("v"))
                    return u.searchParams.get("v");
                const path = u.pathname;
                if (path.startsWith("/shorts/"))
                    return path.replace("/shorts/", "");
                if (path.startsWith("/embed/"))
                    return path.replace("/embed/", "");
            }
            return null;
        }
        catch {
            return null;
        }
    }
    extractListId(input) {
        try {
            const u = new URL(input);
            return u.searchParams.get("list");
        }
        catch {
            return null;
        }
    }
}
exports.YouTubePlugin = YouTubePlugin;
function toSeconds(d) {
    if (typeof d === "number")
        return d;
    if (typeof d === "string") {
        // mm:ss or hh:mm:ss
        const parts = d.split(":").map(Number);
        if (parts.some((n) => Number.isNaN(n)))
            return undefined;
        if (parts.length === 3)
            return parts[0] * 3600 + parts[1] * 60 + parts[2];
        if (parts.length === 2)
            return parts[0] * 60 + parts[1];
        const asNum = Number(d);
        return Number.isFinite(asNum) ? asNum : undefined;
    }
    if (d && typeof d === "object") {
        if (typeof d.seconds === "number")
            return d.seconds;
        if (typeof d.milliseconds === "number")
            return Math.floor(d.milliseconds / 1000);
    }
    return undefined;
}
