// @ziplayer/extension/dist/handlers/WebSocketHandler.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebSocketHandler = void 0;
const ws_1 = __importDefault(require("ws"));
class WebSocketHandler {
    constructor(debug) {
        this.eventCallbacks = new Map();
        this.debug = (message, ...optional) => {
            if (!debug)
                return;
            const formatted = `[WebSocketHandler] ${message}`;
            console.log(formatted, ...optional);
        };
    }
    async connectWebSocket(node, userId, clientName) {
        if (!userId) {
            throw new Error("User ID is required for WebSocket connection");
        }
        const secure = node.secure ?? true;
        const port = node.port ?? (secure ? 443 : 2333);
        const wsProtocol = secure ? "wss" : "ws";
        const wsURL = `${wsProtocol}://${node.host}:${port}/v4/websocket`;
        const headers = {
            Authorization: node.password,
            "User-Id": userId,
            "Client-Name": clientName,
            ...(node.sessionId && { "Session-Id": node.sessionId }),
        };
        return new Promise((resolve, reject) => {
            const ws = new ws_1.default(wsURL, { headers });
            node.ws = ws;
            ws.on("open", () => {
                this.debug(`WebSocket connected to ${node.identifier}`);
                node.wsConnected = true;
                node.wsReconnectAttempts = 0;
            });
            ws.on("message", (data) => {
                try {
                    const message = JSON.parse(data.toString());
                    this.handleWebSocketMessage(node, message);
                }
                catch (error) {
                    this.debug(`Failed to parse WebSocket message from ${node.identifier}`, error);
                }
            });
            ws.on("close", (code, reason) => {
                this.debug(`WebSocket closed for ${node.identifier}: ${code} ${reason.toString()}`);
                node.wsConnected = false;
                node.ws = undefined;
                // Auto-reconnect if not manually closed
                if (code !== 1000 && node.wsReconnectAttempts < node.maxReconnectAttempts) {
                    node.wsReconnectAttempts++;
                    this.debug(`Attempting to reconnect WebSocket for ${node.identifier} (${node.wsReconnectAttempts}/${node.maxReconnectAttempts})`);
                    setTimeout(() => {
                        this.connectWebSocket(node, userId, clientName).catch((error) => this.debug(`WebSocket reconnection failed for ${node.identifier}`, error));
                    }, 5000 * node.wsReconnectAttempts);
                }
            });
            ws.on("error", (error) => {
                this.debug(`WebSocket error for ${node.identifier}`, error);
                node.wsConnected = false;
                reject(error);
            });
            // Resolve after ready event is received
            const originalHandleMessage = this.handleWebSocketMessage.bind(this);
            this.handleWebSocketMessage = (node, message) => {
                if (message.op === "ready") {
                    resolve();
                }
                originalHandleMessage(node, message);
            };
        });
    }
    handleWebSocketMessage(node, message) {
        switch (message.op) {
            case "ready": {
                const readyMsg = message;
                node.sessionId = readyMsg.sessionId;
                this.debug(`Node ${node.identifier} session ready: ${readyMsg.sessionId} (resumed: ${readyMsg.resumed})`);
                break;
            }
            case "stats": {
                const statsMsg = message;
                node.stats = {
                    players: statsMsg.players,
                    playingPlayers: statsMsg.playingPlayers,
                    uptime: statsMsg.uptime,
                    memory: statsMsg.memory,
                    cpu: statsMsg.cpu,
                    frameStats: statsMsg.frameStats,
                };
                this.debug(`Node ${node.identifier} stats updated`, node.stats);
                break;
            }
            case "playerUpdate": {
                const playerUpdateMsg = message;
                this.handlePlayerUpdate(node, playerUpdateMsg);
                break;
            }
            case "event": {
                const eventMsg = message;
                this.handleLavalinkEvent(node, eventMsg);
                break;
            }
            default:
                this.debug(`Unknown WebSocket message type: ${message.op}`);
        }
    }
    handlePlayerUpdate(node, message) {
        this.debug(`Player update for guild ${message.guildId} on node ${node.identifier}`);
        this.emit("playerUpdate", node, message);
    }
    handleLavalinkEvent(node, message) {
        this.debug(`Lavalink event ${message.type} for guild ${message.guildId} on node ${node.identifier}`);
        this.emit("event", node, message);
        this.emit(message.type, node, message);
    }
    closeWebSocket(node) {
        if (node.ws) {
            // Close with a different code to indicate graceful shutdown
            node.ws.close(1001, "Extension destroyed");
            node.ws = undefined;
            node.wsConnected = false;
        }
    }
    closeAllWebSockets(nodes) {
        for (const node of nodes) {
            this.closeWebSocket(node);
        }
    }
    // Event callback system
    on(event, callback) {
        if (!this.eventCallbacks.has(event)) {
            this.eventCallbacks.set(event, []);
        }
        this.eventCallbacks.get(event).push(callback);
    }
    off(event, callback) {
        const callbacks = this.eventCallbacks.get(event);
        if (callbacks) {
            const index = callbacks.indexOf(callback);
            if (index > -1) {
                callbacks.splice(index, 1);
            }
        }
    }
    emit(event, node, data) {
        const callbacks = this.eventCallbacks.get(event);
        if (callbacks) {
            for (const callback of callbacks) {
                try {
                    callback(node, data);
                }
                catch (error) {
                    this.debug(`Error in event callback for ${event}`, error);
                }
            }
        }
    }
}
exports.WebSocketHandler = WebSocketHandler;
